src/board.rs

```diff
@@ -80,12 +80,38 @@ impl Board {
         }
     }
 
-    pub fn move_piece(&mut self, from: Position, to: Position) -> ChessResult<()> {
+    pub fn move_piece<F>(
+        &mut self,
+        from: Position,
+        to: Position,
+        promotion_callback: F,
```

Comment:
So I think we shouldn't necessarily need to pass a `promotion_callback` anytime we make a move, regardless of if we are promoting. Maybe there could be a `get_promotion_callback`, that would ultimately ask the user, or could be set to always queen by the user as well.